### PR TITLE
#4882 : Removed the TestIdentifier.SerializedForm

### PR DESCRIPTION
**Pull request by mehulimukherjee**
**This PR removes the old, custom Java serialization support for TestIdentifier. The SerializedForm inner class has been deleted, and TestIdentifier no longer implements Serializable.**

**What changed**

1. Deleted the SerializedForm inner class from TestIdentifier.
2. Removed implements Serializable and related code.
3. Updated tests:
   - Removed old round-trip serialization tests.
   - Added behavior-based tests to check IDs, names, tags, type, and parent linkage.
   - Added a negative test to confirm that trying to serialize a TestIdentifier now fails with NotSerializableException.
   
TestIdentifier is no longer serializable. If you were persisting or sending these objects via Java serialization, you’ll need to switch to using stable identifiers (getUniqueId()) or create your own small DTO with just the fields you need.
Instead of writing out the whole TestIdentifier, do something like this:

```
// Before (JUnit 5.x) - not supported anymore in JUnit 6
// oos.writeObject(testIdentifier);

// After - recommended
String id = testIdentifier.getUniqueId();
// send/store this string instead
```

**Open questions:**
1. Should I add a short note in the Javadoc for TestIdentifier saying “not serializable”?

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit-framework/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Method [preconditions](https://docs.junit.org/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] [Coding conventions](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://docs.junit.org/snapshot/user-guide/) and [Release Notes](https://docs.junit.org/snapshot/user-guide/#release-notes)

### CODE REVIEW

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/TestIdentifier.java [hunk 1]
```diff
@@ -47,14 +40,7 @@
...
 @API(status = STABLE, since = "1.0")
-public final class TestIdentifier implements Serializable {
-
-	@Serial
-	private static final long serialVersionUID = 1L;
-	@Serial
-	@SuppressWarnings("UnusedVariable")
-	private static final ObjectStreamField[] serialPersistentFields = ObjectStreamClass.lookup(
-		SerializedForm.class).getFields();
+public final class TestIdentifier {
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/TestIdentifier.java]:
```suggestion
public final class TestIdentifier implements Serializable {

	@Serial
	private static final long serialVersionUID = 2L;
```

File: platform-tests/src/test/java/org/junit/platform/launcher/TestIdentifierTests.java [hunk 2]
```diff
@@ -57,34 +60,102 @@ void inheritsTypeFromDescriptor() {
...
+		var engineDescriptor = new EngineDescriptor(UniqueId.forEngine("engine"), "Engine");
+		engineDescriptor.addChild(recreatedDescriptor);
+
+		var rebuiltIdentifier = TestIdentifier.from(recreatedDescriptor);
+		assertDeepEquals(originalIdentifier, rebuiltIdentifier);
 	}
 
 	@Test
-	void initialVersionCanBeDeserialized() throws Exception {
-		try (var inputStream = getClass().getResourceAsStream("/serialized-test-identifier")) {
```
Comment (marcphilipp) [inline platform-tests/src/test/java/org/junit/platform/launcher/TestIdentifierTests.java:68]: I think we can remove this test and the corresponding file (`platform-tests/src/test/resources/serialized-test-identifier`).

**Review by marcphilipp — CHANGES_REQUESTED** [review]

**Comment by mehulimukherjee** [comment]
Hi @marcphilipp ,
Making the changes as you suggested.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/TestIdentifier.java [hunk 3]
```diff
@@ -50,10 +42,6 @@ public final class TestIdentifier implements Serializable {
 
 	@Serial
 	private static final long serialVersionUID = 1L;
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/TestIdentifier.java]: This should be incremented to indicate to Java that the pre-6.0 version of this class and this version are not compatible regarding serialization.

```suggestion
	private static final long serialVersionUID = 2L;
```

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/TestIdentifier.java [hunk 4]
```diff
@@ -50,10 +42,6 @@ public final class TestIdentifier implements Serializable {
...
-	@Serial
-	@SuppressWarnings("UnusedVariable")
-	private static final ObjectStreamField[] serialPersistentFields = ObjectStreamClass.lookup(
-		SerializedForm.class).getFields();
 
 	// These are effectively final but not technically due to late initialization when deserializing
 	private /* final */ UniqueId uniqueId;
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/TestIdentifier.java]: This is no longer the case. Please remove this comment and convert `/* final */` to `final` on all fields.

```suggestion
	private final UniqueId uniqueId;
```

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/TestIdentifier.java [hunk 5]
```diff
@@ -90,19 +78,11 @@ private TestIdentifier(UniqueId uniqueId, String displayName, @Nullable TestSour
 		this.parentId = parentId;
 		this.displayName = displayName;
 		this.source = source;
-		this.tags = copyOf(tags);
+		this.tags = new LinkedHashSet<>(tags);
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/TestIdentifier.java]: Please revert this particular change and restore the `copyOf()` method. Since most descriptors don't have tags it's a non-negligible optimization.

File: platform-tests/src/test/java/org/junit/platform/launcher/TestIdentifierTests.java [hunk 6]
```diff
@@ -63,15 +63,6 @@ void currentVersionCanBeSerializedAndDeserialized() throws Exception {
 		assertDeepEquals(originalIdentifier, deserializedIdentifier);
 	}
```
Comment (marcphilipp) [inline platform-tests/src/test/java/org/junit/platform/launcher/TestIdentifierTests.java:101]: I think we should parameterize this test with the number of tags so serialization is tested with 0, 1, and 2 tags to ensure `copyOf()` actually always returns `Serializable` implementations of `Set`. I can do it as a follow-up, though.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

**Comment by marcphilipp** [comment]
Thank you for your contribution, @mehulimukherjee! :+1:
